evalml/pipelines/components/transformers/imputers/knn_imputer.py

```diff
@@ -0,0 +1,158 @@
+
+        imputer = Sk_KNNImputer(
+            n_neighbors=number_neighbors,
+            missing_values=np.nan,
+            **kwargs,
+        )
+        self._all_null_cols = None
+        super().__init__(
+            parameters=parameters,
+            component_obj=imputer,
+            random_seed=random_seed,
+        )
+
+    def _drop_natural_language_columns(self, X):
+        natural_language_columns = list(
+            X.ww.select(["NaturalLanguage"], return_schema=True).columns.keys(),
+        )
+        if natural_language_columns:
+            X = X.ww.copy()
+            X = X.ww.drop(columns=natural_language_columns)
+        return X, natural_language_columns
+
+    def _set_boolean_columns_to_categorical(self, X):
+        boolean_columns = list(
+            X.ww.select(["Boolean"], return_schema=True).columns.keys(),
+        )
+        if boolean_columns:
+            X = X.ww.copy()
+            X.ww.set_types({col: "Categorical" for col in boolean_columns})
+        return X
+
+    def fit(self, X, y=None):
+        """Fits imputer to data. 'None' values are converted to np.nan before imputation and are treated as the same.
+
+        Args:
+            X (pd.DataFrame or np.ndarray): the input training data of shape [n_samples, n_features]
+            y (pd.Series, optional): the target training data of length [n_samples]
+
+        Returns:
+            self
+
+        Raises:
+            ValueError: if the KNNImputer receives a dataframe with both Boolean and Categorical data.
+
+        """
+        X = infer_feature_types(X)
+
+        if set([lt.type_string for lt in X.ww.logical_types.values()]) == {
+            "boolean",
+            "categorical",
+        } and not all(
+            [
+                is_categorical_actually_boolean(X, col)
+                for col in X.ww.select("Categorical")
+            ],
+        ):
+            raise ValueError(
+                "KNNImputer cannot handle dataframes with both boolean and categorical features.  Use Imputer instead.",
+            )
+
+        nan_ratio = X.ww.describe().loc["nan_count"] / X.shape[0]
+        self._all_null_cols = nan_ratio[nan_ratio == 1].index.tolist()
+
+        X, _ = self._drop_natural_language_columns(X)
+        X = self._set_boolean_columns_to_categorical(X)
+
+        # If the Dataframe only had natural language columns, do nothing.
+        if X.shape[1] == 0:
+            return self
+
+        self._component_obj.fit(X, y)
+        return self
+
+    def transform(self, X, y=None):
+        """Transforms input by imputing missing values. 'None' and np.nan values are treated as the same.
+
+        Args:
+            X (pd.DataFrame): Data to transform.
+            y (pd.Series, optional): Ignored.
+
+        Returns:
+            pd.DataFrame: Transformed X
+        """
+        X = infer_feature_types(X)
+
+        # Return early since bool dtype doesn't support nans and sklearn errors if all cols are bool
```

Comment:
if the columns are boolean they would be of dtype `boolean` instead of `bool` as that is what pandas uses for boolean nullables so it doesn't seem like we would hit this case. Can you explain why sklearn errors if all cols are bool? Related to my comment above, shouldn't we be transforming these columns to categorical?